File: coinphenewalletservice/src/utils/encryption.ts
import { createCipheriv, createDecipheriv, CipherKey, BinaryLike } from "crypto";
import config from "../config";

function _generateRandomIV() {
  let code = '';
  let possible = '0123456789';

  for (let i = 0; i < 16; i += 1) {
    code += possible.charAt(Math.floor(Math.random() * possible.length));
  }

  return code;
}

export function encrypt(str: string) {
  let randomIV = _generateRandomIV();
  let iv = Buffer.from(randomIV);
  const data = Buffer.from(str);

  const cipher = createCipheriv('aes-256-ctr', config.encryptionKey as CipherKey, iv);
  let encrypted = Buffer.concat([cipher.update(data), cipher.final()]);

  let encryptedHex = encrypted.toString('hex');
  let ivHex = iv.toString('hex');

  return encryptedHex.slice(0, 10) + ivHex + encryptedHex.slice(10, encryptedHex.length);
}

export function decrypt(hash: string) {
  const iv = hash.slice(10, 42);
  const content = hash.slice(0, 10) + hash.slice(42, hash.length);

  const decipher = createDecipheriv('aes-256-ctr', config.encryptionKey as BinaryLike, Buffer.from(iv, 'hex'));
  const decrypted = Buffer.concat([decipher.update(Buffer.from(content, 'hex')), decipher.final()]);

  return decrypted.toString();
}

File: coinphenebot/common/encryption.py
import os
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
import config


encryption_key = config.encryption_key.encode("utf-8")  # Must be 32 bytes for AES-256

def _generate_random_iv():
    possible = '0123456789'
    return ''.join([possible[os.urandom(1)[0] % len(possible)] for _ in range(16)])

def encrypt(data: str):
    random_iv = _generate_random_iv()
    iv = bytes(random_iv, 'utf-8')
    
    cipher = Cipher(algorithms.AES(encryption_key), modes.CTR(iv), backend=default_backend())
    encryptor = cipher.encryptor()

    encrypted = encryptor.update(data.encode('utf-8')) + encryptor.finalize()

    encrypted_hex = encrypted.hex()
    iv_hex = iv.hex()

    return encrypted_hex[:10] + iv_hex + encrypted_hex[10:]

def decrypt(hash_str: str):
    iv = bytes.fromhex(hash_str[10:42])
    content = hash_str[:10] + hash_str[42:]

    cipher = Cipher(algorithms.AES(encryption_key), modes.CTR(iv), backend=default_backend())
    decryptor = cipher.decryptor()

    decrypted = decryptor.update(bytes.fromhex(content)) + decryptor.finalize()

    return decrypted.decode('utf-8')
